File: core/src/engine/core/objects/components/Transform.cpp
#include "engine/core/objects/components/Transform.h"
#include "engine/core/objects/components/Script.h"

Transform::Transform() : position(0.0f, 0.0f) {}

Transform::~Transform() {}

void Transform::SetPosition(float x, float y) {
    position.x = x;
    position.y = y;
}

Vector2 Transform::GetPosition() const {
    return position;
}


// Internal API [KEngine.Components.Transform]
void Transform::Internal_GetPosition(void* nativeTransformPtr, Vector2* outPosition) {
    Transform* transform = (Transform*)nativeTransformPtr;
    if (transform) {
        *outPosition = transform->GetPosition();
    }
}

void Transform::Internal_SetPosition(void* nativeTransformPtr, Vector2* position) {
    Transform* transform = (Transform*)nativeTransformPtr;
    if (transform) {
        transform->SetPosition(position->x, position->y);
    }
}

File: KEngine.NET.SDK/Components/Transform.cs
using Kengine.Components;
using System;
using System.Runtime.CompilerServices;

namespace KEngine.Components
{
    public class Transform
    {
        private IntPtr nativeTransformPtr;

        public Transform(IntPtr nativeTransformPtr)
        {
            this.nativeTransformPtr = nativeTransformPtr;
        }

        public Vector2 Position
        {
            get
            {
                Vector2 position;
                Internal_GetPosition(nativeTransformPtr, out position);
                return position;
            }
            set
            {
                Internal_SetPosition(nativeTransformPtr, ref value);
            }
        }

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void Internal_GetPosition(IntPtr nativeTransformPtr, out Vector2 position);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void Internal_SetPosition(IntPtr nativeTransformPtr, ref Vector2 position);
    }
}
